pageserver/src/walreceiver/walreceiver_connection.rs

```diff
@@ -302,21 +313,23 @@ pub async fn handle_walreceiver_connection(
             }
         }
 
-        timeline.check_checkpoint_distance().with_context(|| {
+        let timeline_read = acquire_timeline_read(id, &timeline)?;
```

Comment:
We're still in the loop. Shouldn't we `Arc::clone` the `walingest_timeline` that we got earlier to avoid bailing out at this point?

Generally, we could move the loop body into a separate function and pass in the `Arc<Timeline>` from the caller. That ensures that the loop body always has an alive timeline at hand.

Reply:
Nice catch, replaced all `timeline_read` with the `walingest_timeline`.
Yet, I've not dared to move away entire thing, since it requires a lot of changes: the loop needs quite a lot of parameters to operate.